cfme/tests/automate/custom_button/test_cloud_objects.py

```diff
@@ -303,3 +303,65 @@ def test_custom_button_automate(appliance, request, submit, setup_objs, button_g
                 assert False, "Expected {} requests not found in automation log".format(
                     str(expected_count)
                 )
+
+
+@pytest.mark.uncollectif(
+    lambda appliance, button_group: not bool([obj for obj in OBJ_TYPE_59 if obj in button_group])
+    and appliance.version < "5.10"
+)
+@pytest.mark.parametrize("expression", ["enablement", "visibility"])
+def test_custom_button_expression(appliance, request, setup_objs, button_group, expression):
+    """ Test custom button as per expression enablement/visibility.
+
+    prerequisites:
+        * Appliance with Cloud Provider
+
+    Steps:
+        * Create custom button group with the Object type
+        * Create a custom button with expression (Tag)
+            1. Enablement Expression
+            2. Visibility Expression
+        * Navigate to object Detail page
+        * Check: button should not enable/visible without tag
+        * Check: button should enable/visible with tag
+    """
+
+    group, obj_type = button_group
+    exp = {expression: {"tag": "My Company Tags : Department", "value": "Engineering"}}
+    button = group.buttons.create(
+        text=fauxfactory.gen_alphanumeric(),
+        hover=fauxfactory.gen_alphanumeric(),
+        display_for="Single entity",
+        system="Request",
+        request="InspectMe",
+        **exp
+    )
+    request.addfinalizer(button.delete_if_exists)
+
+    tag_cat = appliance.collections.categories.instantiate(
+        name="department", display_name="Department"
+    )
+    tag = tag_cat.collections.tags.instantiate(name="engineering", display_name="Engineering")
```

Comment:
The test up to here seems to be duplicated across many of these tests, and would likely be better as a fixture with some indirect parametrization.  (TODO)